File: demo/java/src/main/java/Demo1.java
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.HashMap;
import java.util.Map;

public class Demo1 {
    public static void main(String[] args) {
        Company company = new Company();
        CompanyBuilder builder = company.getBuilder();
        builder.setId(1);
        builder.setCode(1111);
        builder.setName("coco");
        builder.setAddr("aaaaaaaa");

        byte[] data = builder.toByteArray();

        Company company1 = new Company();
        Company cpy = company1.parseFrom(data);
        cpy.getId();
        cpy.getCode();
    }
}

class CoType
{
    public static final int CO_INT = 1;
    public static final int CO_LONG = 2;
    public static final int CO_STRING = 3;
}

class Util
{
    public static byte[] int2Byte(int val) {
        ByteBuffer buf = ByteBuffer.allocate(4);
        buf.order(ByteOrder.BIG_ENDIAN);
        buf.putInt(val);
        return buf.array();
    }

    public static byte[] long2Byte(long val) {
        ByteBuffer buf = ByteBuffer.allocate(8);
        buf.order(ByteOrder.BIG_ENDIAN);
        buf.putLong(val);
        return buf.array();
    }

    public static byte[] byteMerge(byte[] b1, byte[] b2) {
        if (b1 == null) {
            return b2;
        }

        if (b2 == null) {
            return b1;
        }

        byte[] b3 = new byte[b1.length + b2.length];
        System.arraycopy(b1, 0, b3, 0, b1.length);
        System.arraycopy(b2, 0, b3, b1.length, b2.length);
        return b3;
    }
}
class DataType
{
    protected byte[] byteVal = null;

    public void setVal(Object val)
    {

    }

    public byte[] packByteVal(int tag, int type, boolean need) {
        if(this.byteVal != null) {
            if (tag >= 15) {
                byte[] ret = new byte[this.byteVal.length + 2];
                ret[0] = (byte)((type << 4) | 15);
                ret[1] = (byte)(tag);
                for (int i = 2; i < ret.length; i++)
                {
                    ret[i] = this.byteVal[i - 2];
                }
                return ret;
            } else {
                byte[] ret = new byte[this.byteVal.length + 1];
                ret[0] = (byte)((type << 4) | tag);
                for (int i = 1; i < ret.length; i++)
                {
                    ret[i] = this.byteVal[i - 1];
                }
                return ret;
            }
        }

        return null;
    }

    public void setByteVal(Object val, int type) {
        if (type == CoType.CO_INT) {
            this.byteVal = Util.int2Byte((int) val);
        } else if (type == CoType.CO_LONG) {
            this.byteVal = Util.long2Byte((long) val);
        } else if (type == CoType.CO_STRING) {
            String str = (String) val;
            if (str == null) {
                this.byteVal = null;
            } else {
                byte[] strBytes = str.getBytes();
                this.byteVal = Util.byteMerge(Util.int2Byte(strBytes.length), strBytes);
            }
        }
    }
}

class CoInt extends DataType
{
    public int val;

    public void setVal(int val)
    {
        this.val = val;
    }

    public byte[] toByteVal(int tag, boolean need) {
        this.setByteVal(this.val, CoType.CO_INT);
        return this.packByteVal(tag, CoType.CO_INT, need);
    }

    public int getVal() {
        return val;
    }

    public CoInt parseByteVal(Parser parser, int tag) {
        this.val = (int) parser.getVal(tag);
        return this;
    }
}

class CoLong extends DataType
{
    public long val;

    public void setVal(long val)
    {
        this.val = val;
    }

    public long getVal() {
        return val;
    }

    public byte[] toByteVal(int tag, boolean need) {
        this.setByteVal(this.val, CoType.CO_LONG);
        return this.packByteVal(tag, CoType.CO_LONG, need);
    }

    public CoLong parseByteVal(Parser parser, int tag) {
        this.val = (long) parser.getVal(tag);
        return this;
    }
}

class CoString extends DataType {
    public String val;

    public void setVal(String val)
    {
        this.val = val;
    }

    public String getVal() {
        return val;
    }

    public byte[] toByteVal(int tag, boolean need) {
        this.setByteVal(this.val, CoType.CO_STRING);
        return this.packByteVal(tag, CoType.CO_STRING, need);
    }

    public CoString parseByteVal(Parser parser, int tag) {
        this.val = (String) parser.getVal(tag);
        return this;
    }
}

class Company {
    public int id;
    public long code;
    public String name;
    public String addr;
    protected CompanyBuilder builder;

    Company() {
        this.builder = new CompanyBuilder();
    }

    public void setId(int id) {
        this.id = id;
    }

    public void setCode(long code) {
        this.code = code;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setAddr(String addr) {
        this.addr = addr;
    }

    public int getId() {
        return id;
    }

    public long getCode() {
        return code;
    }

    public String getName() {
        return name;
    }

    public String getAddr() {
        return addr;
    }

    public CompanyBuilder getBuilder() {
        return builder;
    }

    public Company parseFrom(byte[] data) {
        this.builder.parse(data, this);
        return this;
    }
}

class CompanyBuilder {
    public CoInt id;
    public CoLong code;
    public CoString name;
    public CoString addr;

    CompanyBuilder()
    {
        this.id = new CoInt();
        this.code = new CoLong();
        this.name = new CoString();
        this.addr = new CoString();
    }

    public void setId(int id)
    {
        this.id.setVal(id);
    }

    public void setCode(long code)
    {
        this.code.setVal(code);
    }

    public void setName(String name)
    {
        this.name.setVal(name);
    }

    public void setAddr(String addr)
    {
        this.addr.setVal(addr);
    }

    public byte[] toByteArray() {
        byte[] ret = this.id.toByteVal(18, true);
        ret = Util.byteMerge(ret, this.code.toByteVal(2, true));
        ret = Util.byteMerge(ret, this.name.toByteVal(3, false));
        ret = Util.byteMerge(ret, this.addr.toByteVal(4, false));
        return ret;
    }

    public void parse(byte[] data, Company company) {
        Parser parser = new Parser(data);
        this.id.parseByteVal(parser, 18);
        this.code.parseByteVal(parser, 2);
        this.name.parseByteVal(parser, 3);
        this.addr.parseByteVal(parser, 4);

        company.setId(this.id.getVal());
        company.setCode(this.code.getVal());
        company.setName(this.name.getVal());
        company.setAddr(this.addr.getVal());
    }
}

class Parser {
    public Map parserData;

    Parser(byte[] data) {
        ByteBuffer buf = ByteBuffer.allocate(data.length);
        buf.put(data);

        parserData = new HashMap();

        int offset = 0;
        while (offset < data.length) {
            int typeAndTag = (int) buf.get(offset);
            offset++;

            int type = typeAndTag >> 4;
            int tag = typeAndTag - (type << 4);

            if (tag == 15) {
                tag = (int) buf.get(offset);
                offset++;
            }

            if (type == CoType.CO_INT) {
                parserData.put(tag, buf.getInt(offset));
                offset = offset + 4;
            } else if (type == CoType.CO_LONG) {
                parserData.put(tag, buf.getLong(offset));
                offset = offset + 8;
            } else if (type == CoType.CO_STRING) {
                int len = buf.getInt(offset);
                offset = offset + 4;

                byte[] str = new byte[len];
                for (int i = 0; i < len; i++) {
                    str[i] = buf.get(offset);
                    offset++;
                }

                parserData.put(tag, new String(str));
            }
        }
    }

    public Object getVal(int tag) {
        if (parserData.containsKey(tag)) {
            return parserData.get(tag);
        } else {
            return null;
        }
    }
}

File: demo/php/demo1.php
<?php 
// struct Company
// {
//     0 need int id;
//     1 need long code;
//     2 need string name;
//     3 optional string addr;
// }
// 

class Company {
    public $id;
    public $code;
    public $name;
    public $addr;

    public function __construct()
    {
        $this->id = new CoInt;
        $this->code = new CoLong;
        $this->name = new CoString;
        $this->addr = new CoString;
    }

    public function setId($id)
    {
        $this->id->setVal($id);
    }

    public function setCode($code)
    {
        $this->code->setVal($code);
    }

    public function setName($name)
    {
        $this->name->setVal($name);
    }

    public function setAddr($addr)
    {
        $this->addr->setVal($addr);
    }

    public function pack(&$val)
    {
        $this->id->get($val, 0, true);
        $this->code->get($val, 1, true);
        $this->name->get($val, 2, true);
        $this->addr->get($val, 3, false);
    }

    public function unpack($data)
    {   
        $this->id = $this->id->getUnpackData($data, 0);
        $this->code = $this->code->getUnpackData($data, 1);
        $this->name = $this->name->getUnpackData($data, 2);
        $this->addr = $this->addr->getUnpackData($data, 3);
    }
}

$data = null;
$obj = new Company;
$obj->setId(10);
$obj->setCode(1111);
$obj->setName('coco');
$obj->setAddr('aaaaaaaaaa');
$obj->pack($data);

echo "压缩后总长度为".strlen($data)."\n";
$company = new Company;
$company->unpack(coUnpack($data));
print_r($company);

$data = '{"id":10,"code":1111,"name":"coco","addr":"aaaaaaaaaa"}';
echo $data.'总长度为'.strlen($data)."\n";
//总结就是如果字段越多，自定义协议的传输数据的优势越明显。


//lib
function coUnpack($data)
{
    $len = 0;
    $unpackData = [];
    while (strlen($data) - $len > 0) {

        if ($len > 0) $data = substr($data, $len);
        $type = ord(substr($data, 0, 1));

        $data = substr($data, 1);
        $tag = ord(substr($data, 0, 1));

        $data = substr($data, 1);

        if ($type == CoType::CO_INT) {
            $len = 4;
            $val = unpack("N", substr($data, 0, 4));
            $val = $val[1];
        } elseif ($type == CoType::CO_LONG) {
            $len = 8;
            $val = unpack("J", substr($data, 0, 8));
            $val = $val[1];
        } else  {
            $len = unpack("N", substr($data, 0, 4));
            $len = $len[1];

            $data = substr($data, 4);
            $val = unpack("a*", substr($data, 0, $len));
            $val = $val[1];
        }

        $unpackData[$tag] = [
            'type' => $type,
            'tag' => $tag,
            'len' => $len,
            'val' => $val,
        ];
    }

    return $unpackData;
}

function coPack($data, $type, $tag, &$val)
{   
    $val .= chr($type).chr($tag).$data;
}

class CoType
{
    const CO_INT = 1;
    const CO_LONG = 2;
    const CO_STRING = 3;
}

class DataType 
{   
    public function packVal(&$val, $type, $tag, $isNeed = true)
    {   
        if (!isset($this->val)) {
            $this->val = null;
        }

        if ($this->val != null) {
            coPack($this->val, $type, $tag, $val);
        } else {
            if ($isNeed) {
                throw new Exception("缺少必要的参数标记为:{$tag}", 1);
            }
        }
    }

    public function setVal($data)
    {
        $this->val = pack("a*", $data);
        $bodyLen = strlen($this->val);
        $head = pack("N", $bodyLen);
        $this->val = $head.$this->val;
        return $this->val;
    }
}

class CoInt extends DataType
{   
    public function setVal($data)
    {   
        $this->val = pack("N", intval($data));
        return $this->val;
    }

    public function get(&$val, $tag, $isNeed = true)
    {
        $this->packVal($val, CoType::CO_INT, $tag, $isNeed);
    }

    public function getUnpackData($data, $tag)
    {
        if (isset($data[$tag]) && $data[$tag]['type'] == CoType::CO_INT) {
            return $data[$tag]['val'];
        }
    }
}

class CoLong extends DataType
{   
    public function setVal($data)
    {   
        $this->val = pack("J", intval($data));
        return $this->val;
    }

    public function get(&$val, $tag, $isNeed = true)
    {
        $this->packVal($val, CoType::CO_LONG, $tag, $isNeed);
    }

    public function getUnpackData($data, $tag)
    {
        if (isset($data[$tag]) && $data[$tag]['type'] == CoType::CO_LONG) {
            return $data[$tag]['val'];
        }
    }
}

class CoString extends DataType
{   
    public function get(&$val, $tag, $isNeed = true)
    {
        $this->packVal($val, CoType::CO_STRING, $tag, $isNeed);
    }

    public function getUnpackData($data, $tag)
    {
        if (isset($data[$tag]) && $data[$tag]['type'] == CoType::CO_STRING) {
            return $data[$tag]['val'];
        }
    }
}

